[SYCL][Fusion] Scheduler support for kernel fusion

sommerlukas opened the pull request:
This is the third patch in a series of patches to add an implementation of the [kernel fusion extension](https://github.com/intel/llvm/pull/7098). We have split the implementation into multiple patches to make them more easy to review. This patch integrates the kernel fusion extension into the SYCL runtime scheduler. 

Next to collecting the kernels submitted while in fusion mode in the fusion list associated with the queue, the integration into the scheduler is also responsible for detecting the synchronization scenarios. Various scenarios, such as buffer destruction or event wait, require fusion to be aborted early. The full list of scenarios is available in the [extension proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_codeplay_kernel_fusion.asciidoc#synchronization-in-the-sycl-application).

A high-level description of the integration into the scheduler can be found in the [design document](https://github.com/intel/llvm/pull/7204). 

This PR can be reviewed and merged independently of https://github.com/intel/llvm/pull/7465.

Signed-off-by: Lukas Sommer <lukas.sommer@codeplay.com>

sommerlukas commented:
Tests for this PR are in https://github.com/intel/llvm-test-suite/pull/1416

dm-vodopyanov commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1416

steffenlarsen commented on sycl/source/detail/fusion/fusion_wrapper_impl.cpp:
```diff
@@ -20,22 +22,23 @@ std::shared_ptr<detail::queue_impl> fusion_wrapper_impl::get_queue() const {
 
-bool fusion_wrapper_impl::is_in_fusion_mode() const { return false; }
+bool fusion_wrapper_impl::is_in_fusion_mode() const {
+  auto QUniqueID = std::hash<QueueImplPtr>()(MQueue);
+  return detail::Scheduler::getInstance().isInFusionMode(QUniqueID);
```
Alternatively:
```suggestion
  return MQueue->is_in_fusion_mode();
```
Not a huge difference, but a little less code to maintain. 😄

steffenlarsen commented on sycl/source/detail/scheduler/graph_builder.cpp:
```diff
@@ -1387,6 +1481,186 @@ Command *Scheduler::GraphBuilder::connectDepEvent(
+    return false;
+  }
+  return FusionList->second->isActive();
+}
```
For the sake of keeping the interface somewhat consistent, would it make sense to have a

```suggestion
}

bool Scheduler::GraphBuilder::isInFusionMode(QueueImplPtr Queue) {
  return isInFusionMode(std::hash<QueueImplPtr>()(Queue));
}
```

and use that in `queue_impl`?

sommerlukas replied:
I see your point about the interface. However, `QueueImplPtr` is `std::shared_ptr<detail::queue_impl>`, so in order to be able to call this from `queue_impl`, we would need to create a new shared pointer from `this` in `queue_impl`, which would lose the reference count.

steffenlarsen replied:
Good point. Let's keep it as-is.

steffenlarsen reviewed (commented):
Generally looks good, but since it touches the graph-builder and scheduler I would like @sergey-semenov's eyes on it too.

bader commented:
> Generally looks good, but since it touches the graph-builder and scheduler I would like @sergey-semenov's eyes on it too.

@sergey-semenov, ping.

sergey-semenov commented on sycl/source/detail/scheduler/graph_builder.cpp:
```diff
@@ -1387,6 +1481,186 @@ Command *Scheduler::GraphBuilder::connectDepEvent(
+        auto *DepReq = PrevDep.MDepRequirement;
+        auto *DepRecord = getMemObjRecord(DepReq->MSYCLMemObj);
+        if (DepRecord == Record) {
+          // Need to restore this as a leave, because we pushed it from the
```
```suggestion
          // Need to restore this as a leaf, because we pushed it from the
```

sergey-semenov commented on sycl/source/detail/scheduler/graph_builder.cpp:
```diff
@@ -1387,6 +1481,186 @@ Command *Scheduler::GraphBuilder::connectDepEvent(
+    Node->MLeafCounter -= Record->MWriteLeaves.remove(Node);
+    // If the placeholder had a write-requirement on this record, we need to
+    // restore the previous leaves.
+    if (AccessMode != access::mode::read) {
```
This looks somewhat worrying. Are we always able to restore the correct leaf commands here? Can there be a scenario where the previous leaf command has already been destroyed at this point (e.g. once it stops being a leaf assuming it's already been enqueued) and we just lose the dependency when adding the fused kernel?

sergey-semenov replied:
I'm not sure, I might be getting something wrong. Let's say there's the following graph (assume all deps have write access): 
`Alloca <- CG`. We then start fusion: `Alloca <- CG <- Fusion list CG A <- Fusion list CG B`, CG is removed from leaves and cleaned up during this process, but isn't finished, so the graph looks like this: `Alloca <- Fusion list CG A <- Fusion list CG B`. Since the event associated with CG has already been added to the list of event dependencies of `Fusion list CG A` when building the graph, we'll honor that dependency if the fusion is cancelled and placeholder CGs are enqueued as normal despite this cleanup.

Would the same hold true if we complete the fusion? We remove the fusion list CGs first while restoring the leaves and then add the fusion cmd to the graph. It seems to me like the only dependency for the fusion command in this scenario would be `Alloca` while `CG` would be lost. Am I missing some way this dependency would be preserved here?

sommerlukas replied:
So far, I assumed that if `CG` gets to the stage where it is cleaned up (destroyed) and wouldn't be available for restoring as a leave through `Fusion list CG A`'s `MDep`, it would already have completed execution. 
 
However, if `CG` can proceed to clean-up when it has been enqueued, but not necessarily executed yet, the scenario you pointed out might indeed be problematic. 

I have pushed a new commit to address this problem by inheriting all dependencies (apart from the ones on other commands in the fusion list) from the input commands in the fusion list. So, in your example, when we call `complete_fusion`, the new command for the fused kernel should inherit the dependency on `CG` from the input command `Fusion list CG A`, so we have the correct dependencies set up.

sommerlukas replied:
Thanks for the feedback @sergey-semenov!

> Could you please add a test for this scenario as well?

This will not be possible as part of this PR. The logic to create a fused kernel from the kernels in the fusion list is only part of a separate, later PR (draft for that PR: https://github.com/sommerlukas/llvm/pull/1). 

Currently, `complete_fusion` always falls back to `cancel_fusion`, see the `TODO` in `complete_fusion` at `graph_builder.cpp:1548`. 

If it is ok with you, I would suggest to merge this PR and I will add a unit test for this scenario to the next PR, which introduces the necessary JIT compilation logic.

sergey-semenov reviewed (commented):
Sorry about the late review. Could you add some unit tests to cover the new functionality?

sommerlukas commented:
@sergey-semenov Thank you very much for your review and feedback! I've addressed it in the latest commit. 

>  Could you add some unit tests to cover the new functionality?

I have now added a unit-test (as part of the `SchedulerTests`), testing correct dependency setup for kernel fusion. 

The scheduler integration is also tested by a number of application tests in https://github.com/intel/llvm-test-suite/pull/1416.

sergey-semenov reviewed (approved):
LGTM

pvchupin commented:
@sommerlukas, please fix conflicts
